internal/controller/clusterorder_webhook.go

```diff
@@ -6,40 +6,100 @@ import (
 	"encoding/json"
 	"fmt"
 	"net/http"
+	"sync"
 	"time"
 
 	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
 
 	cloudkitv1alpha1 "github.com/innabox/cloudkit-operator/api/v1alpha1"
 )
 
-func triggerWebHook(ctx context.Context, url string, instance *cloudkitv1alpha1.ClusterOrder) error {
+type InflightRequest struct {
+	createTime             time.Time
+	minimumRequestInterval time.Duration
+}
+
+var (
+	ifRequestsLock   *sync.RWMutex              = &sync.RWMutex{}
+	inflightRequests map[string]InflightRequest = make(map[string]InflightRequest)
+)
+
+func checkForExistingRequest(url string) time.Duration {
+	var delta time.Duration
+
+	ifRequestsLock.RLock()
+	request, ok := inflightRequests[url]
+	if ok {
+		delta = time.Since(request.createTime)
+		if delta >= request.minimumRequestInterval {
+			delta = 0
+		}
+	}
+	ifRequestsLock.RUnlock()
+	purgeExpiredRequests()
+	return delta
+}
+
+func addInflightRequest(url string, minimumRequestInterval time.Duration) {
+	ifRequestsLock.Lock()
+	inflightRequests[url] = InflightRequest{
+		createTime:             time.Now(),
+		minimumRequestInterval: minimumRequestInterval,
+	}
+	ifRequestsLock.Unlock()
+	purgeExpiredRequests()
+}
+
+func purgeExpiredRequests() {
+	for url, request := range inflightRequests {
+		ifRequestsLock.RLock()
```

Comment:
Again, once you start iterating, you're already reading the map. Here it is too late to get the lock. You need the lock before the `range` statement.

By the time we get to this line, the first key and value have already been read from the map and copied into the values `url` and `request`.

Reply:
Maybe I commented too quickly; did you push a change yet? I saw that the above comment was resolved and assumed this was the current code. Maybe I should be more patient. :)